# <font color="#6e8ebc">**Data Sets**</font>
---

# <font color="#6e8ebc">**Índice**</font>

1. <font color="#6e8ebc">**Configuraciones**</font>
   * Importación de librerías
   * Paths
   * Paletas de colores
2. <font color="#6e8ebc">**Exploración inicial**</font>
   * Dataset Churn Bank
   * Dataset PaySim
3. <font color="#6e8ebc">**Manipulación de datos**</font>
   * Churn
     - Fechas
     - Clusters Iniciales
     - Clusters K-Means
     - Interpretación Clusters K-Means
4. <font color="#6e8ebc">**Transactions - Generación de datos**</font>
   * Chequeos generales
   * Distribuciones numéricas
   * Churn
   * Correlación
5. <font color="#6e8ebc">**App Interaction - Generación de datos**</font>
6. <font color="#6e8ebc">**Guardado Datasets**</font>

# <font color="#6e8ebc">**Cofiguraciones**</font>

## <font color="#ff743d">**Importación de librerías**</font>

In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
import random
import shutil

import kagglehub

from tqdm import tqdm

import os
from pathlib import Path
from typing import Tuple, Callable, List

import warnings
import pickle

In [2]:
warnings.filterwarnings('ignore')
pd.set_option('display.max_columns', None)

RANDOM_SEED = 42

## <font color="#ff743d">**Paths**</font>

In [3]:
# Obtiene el path actual
PROJECT_PATH = os.getcwd()

src = Path('src')
DATA_PATH = PROJECT_PATH / src

img = Path('img')
IMG_PATH = PROJECT_PATH / img

models = Path('models')
MODELS_PATH = PROJECT_PATH / models

reports = Path('reports')
REPORTS_PATH = PROJECT_PATH / reports

api = Path('api')
API_PATH = PROJECT_PATH / api

artifacts = Path('artifacts')
ARTIFACTS_PATH = API_PATH / artifacts

dirs = [DATA_PATH, IMG_PATH, MODELS_PATH, REPORTS_PATH, API_PATH, ARTIFACTS_PATH]


for directory in dirs:
    if not directory.exists():
        directory.mkdir(parents=True)
        print(f'Directorio creado:\n{directory}\n')
    else:
        print(f'El directorio ya existe:\n{directory}\n')

El directorio ya existe:
C:\Users\Ignacio\OneDrive\Documents\git-and-github\hackathon_ONE\src

El directorio ya existe:
C:\Users\Ignacio\OneDrive\Documents\git-and-github\hackathon_ONE\src\monthly_data

El directorio ya existe:
C:\Users\Ignacio\OneDrive\Documents\git-and-github\hackathon_ONE\src\monthly_data\tx

El directorio ya existe:
C:\Users\Ignacio\OneDrive\Documents\git-and-github\hackathon_ONE\src\monthly_data\app

El directorio ya existe:
C:\Users\Ignacio\OneDrive\Documents\git-and-github\hackathon_ONE\img

El directorio ya existe:
C:\Users\Ignacio\OneDrive\Documents\git-and-github\hackathon_ONE\models

El directorio ya existe:
C:\Users\Ignacio\OneDrive\Documents\git-and-github\hackathon_ONE\reports

El directorio ya existe:
C:\Users\Ignacio\OneDrive\Documents\git-and-github\hackathon_ONE\api

El directorio ya existe:
C:\Users\Ignacio\OneDrive\Documents\git-and-github\hackathon_ONE\api\artifacts



## <font color="#ff743d">**Paleta de Colores del Proyecto**</font>

**Naranjas**

<span style="display:inline-block; width:20px; height:20px; background:#993c00; margin-right:5px;"></span>
<span style="display:inline-block; width:20px; height:20px; background:#cc5a00; margin-right:5px;"></span>
<span style="display:inline-block; width:20px; height:20px; background:#ff7b00; margin-right:5px;"></span>
<span style="display:inline-block; width:20px; height:20px; background:#ff9c40; margin-right:5px;"></span>
<span style="display:inline-block; width:20px; height:20px; background:#ffbf80; margin-right:5px;"></span>

**Rojos**

<span style="display:inline-block; width:20px; height:20px; background:#7a1c00; margin-right:5px;"></span>
<span style="display:inline-block; width:20px; height:20px; background:#b23200; margin-right:5px;"></span>
<span style="display:inline-block; width:20px; height:20px; background:#e74903; margin-right:5px;"></span>
<span style="display:inline-block; width:20px; height:20px; background:#ff743d; margin-right:5px;"></span>
<span style="display:inline-block; width:20px; height:20px; background:#ffa68c; margin-right:5px;"></span>

**Azules**

<span style="display:inline-block; width:20px; height:20px; background:#001a37; margin-right:5px;"></span>
<span style="display:inline-block; width:20px; height:20px; background:#002b5c; margin-right:5px;"></span>
<span style="display:inline-block; width:20px; height:20px; background:#003873; margin-right:5px;"></span>
<span style="display:inline-block; width:20px; height:20px; background:#315c99; margin-right:5px;"></span>
<span style="display:inline-block; width:20px; height:20px; background:#6e8ebc; margin-right:5px;"></span>

**Celestes**

<span style="display:inline-block; width:20px; height:20px; background:#004752; margin-right:5px;"></span>
<span style="display:inline-block; width:20px; height:20px; background:#007789; margin-right:5px;"></span>
<span style="display:inline-block; width:20px; height:20px; background:#00b5d0; margin-right:5px;"></span>
<span style="display:inline-block; width:20px; height:20px; background:#40cee6; margin-right:5px;"></span>
<span style="display:inline-block; width:20px; height:20px; background:#8ae6f4; margin-right:5px;"></span>

In [4]:
NARANJA_1, NARANJA_2, NARANJA_3, NARANJA_4, NARANJA_5 = '#993c00', '#cc5a00', '#ff7b00', '#ff9c40', '#ffbf80'
ROJO_1, ROJO_2, ROJO_3, ROJO_4, ROJO_5 = '#7a1c00', '#b23200', '#e74903', '#ff743d', '#ffa68c'
AZUL_1, AZUL_2, AZUL_3, AZUL_4, AZUL_5 = '#001a37', '#002b5c', '#003873', '#315c99', '#6e8ebc'
CIELO_1, CIELO_2, CIELO_3, CIELO_4, CIELO_5 = '#004752', '#007789', '#00b5d0', '#40cee6', '#8ae6f4'

# <font color="#6e8ebc">**Exploración inicial**</font>

## <font color="#ff743d">**Dataset Churn Bank**</font>

In [5]:
def obtener_datos_kagglehub(destination_path):
    """
    Descarga el dataset usando kagglehub y mueve el CSV a la carpeta del proyecto.
    """
    dataset_slug = "mathchi/churn-for-bank-customers"
    
    print(f"⬇️ Solicitando dataset: {dataset_slug} via kagglehub...")
    
    try:
        # 1. Descarga (kagglehub gestiona el caché automáticamente)
        path_cache = kagglehub.dataset_download(dataset_slug)
        print(f"✅ Descargado en caché: {path_cache}")
        
        # 2. Mover el archivo CSV a nuestra carpeta de proyecto (DATA_PATH)
        # Buscamos archivos CSV en la carpeta descargada
        archivos_csv = [f for f in os.listdir(path_cache) if f.endswith('.csv')]
        
        if not archivos_csv:
            print("⚠️ No se encontró ningún archivo CSV en el dataset descargado.")
            return

        for archivo in archivos_csv:
            origen = os.path.join(path_cache, archivo)
            destino = os.path.join(destination_path, archivo)
            
            # Copiamos/Sobreescribimos para asegurar que tenemos la última versión
            shutil.copy(origen, destino)
            print(f"📂 Archivo movido a: {destino}")

    except Exception as e:
        print(f"❌ Error al descargar con kagglehub: {e}")

# --- EJECUCIÓN ---
# Asegurarse de tener instalado: pip install kagglehub
obtener_datos_kagglehub(DATA_PATH)

try:
    csv_path = os.path.join(DATA_PATH, "churn.csv") 
    if os.path.exists(csv_path):
        df_churn = pd.read_csv(csv_path)
        print(f"\n✅ DataFrame cargado con éxito. Shape: {df_churn.shape}")
        print(df_churn.head(3))
except Exception as e:
    print("\n⚠️ El archivo se descargó, pero verifica el nombre exacto en la carpeta data/")

⬇️ Solicitando dataset: mathchi/churn-for-bank-customers via kagglehub...
✅ Descargado en caché: C:\Users\Ignacio\.cache\kagglehub\datasets\mathchi\churn-for-bank-customers\versions\1
📂 Archivo movido a: C:\Users\Ignacio\OneDrive\Documents\git-and-github\hackathon_ONE\src\churn.csv

✅ DataFrame cargado con éxito. Shape: (10000, 14)
   RowNumber  CustomerId   Surname  CreditScore Geography  Gender  Age  \
0          1    15634602  Hargrave          619    France  Female   42   
1          2    15647311      Hill          608     Spain  Female   41   
2          3    15619304      Onio          502    France  Female   42   

   Tenure    Balance  NumOfProducts  HasCrCard  IsActiveMember  \
0       2       0.00              1          1               1   
1       1   83807.86              1          0               1   
2       8  159660.80              3          1               0   

   EstimatedSalary  Exited  
0        101348.88       1  
1        112542.58       0  
2        113931.5

In [6]:
df_churn.head()

,RowNumber,CustomerId,Surname,CreditScore,Geography,Gender,Age,Tenure,Balance,NumOfProducts,HasCrCard,IsActiveMember,EstimatedSalary,Exited
0,1,15634602,Hargrave,619,France,Female,42,2,0.00,1,1,1,101348.88,1
1,2,15647311,Hill,608,Spain,Female,41,1,83807.86,1,0,1,112542.58,0
2,3,15619304,Onio,502,France,Female,42,8,159660.80,3,1,0,113931.57,1
3,4,15701354,Boni,699,France,Female,39,1,0.00,2,0,0,93826.63,0
4,5,15737888,Mitchell,850,Spain,Female,43,2,125510.82,1,1,1,79084.10,0


In [7]:
df_churn.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10000 entries, 0 to 9999
Data columns (total 14 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   RowNumber        10000 non-null  int64  
 1   CustomerId       10000 non-null  int64  
 2   Surname          10000 non-null  object 
 3   CreditScore      10000 non-null  int64  
 4   Geography        10000 non-null  object 
 5   Gender           10000 non-null  object 
 6   Age              10000 non-null  int64  
 7   Tenure           10000 non-null  int64  
 8   Balance          10000 non-null  float64
 9   NumOfProducts    10000 non-null  int64  
 10  HasCrCard        10000 non-null  int64  
 11  IsActiveMember   10000 non-null  int64  
 12  EstimatedSalary  10000 non-null  float64
 13  Exited           10000 non-null  int64  
dtypes: float64(2), int64(9), object(3)
memory usage: 1.1+ MB


In [8]:
df_churn.columns

Index(['RowNumber', 'CustomerId', 'Surname', 'CreditScore', 'Geography',
       'Gender', 'Age', 'Tenure', 'Balance', 'NumOfProducts', 'HasCrCard',
       'IsActiveMember', 'EstimatedSalary', 'Exited'],
      dtype='object')

In [9]:
df_churn.describe()

,RowNumber,CustomerId,CreditScore,Age,Tenure,Balance,NumOfProducts,HasCrCard,IsActiveMember,EstimatedSalary,Exited
count,10000.00000,1.000000e+04,10000.000000,10000.000000,10000.000000,10000.000000,10000.000000,10000.00000,10000.000000,10000.000000,10000.000000
mean,5000.50000,1.569094e+07,650.528800,38.921800,5.012800,76485.889288,1.530200,0.70550,0.515100,100090.239881,0.203700
std,2886.89568,7.193619e+04,96.653299,10.487806,2.892174,62397.405202,0.581654,0.45584,0.499797,57510.492818,0.402769
min,1.00000,1.556570e+07,350.000000,18.000000,0.000000,0.000000,1.000000,0.00000,0.000000,11.580000,0.000000
25%,2500.75000,1.562853e+07,584.000000,32.000000,3.000000,0.000000,1.000000,0.00000,0.000000,51002.110000,0.000000
50%,5000.50000,1.569074e+07,652.000000,37.000000,5.000000,97198.540000,1.000000,1.00000,1.000000,100193.915000,0.000000
75%,7500.25000,1.575323e+07,718.000000,44.000000,7.000000,127644.240000,2.000000,1.00000,1.000000,149388.247500,0.000000
max,10000.00000,1.581569e+07,850.000000,92.000000,10.000000,250898.090000,4.000000,1.00000,1.000000,199992.480000,1.000000


In [10]:
df_churn['Exited'].sum()

2037

In [11]:
df_churn['Exited'].value_counts(normalize=True)

Exited
0    0.7963
1    0.2037
Name: proportion, dtype: float64

In [12]:
df_churn['CustomerId'].nunique()

10000

In [13]:
df_churn['EstimatedSalary'].max()

199992.48

# <font color="#6e8ebc">**Manipulación de datos**</font>

## 🔧 <font color="#ff743d">**Overview de los 3 datasets**</font>

### 1. Churn (base)
- Target + atributos estáticos.

### 2. Transacciones
- Churners:
    * Clientes que muestran una caída de actividad en los últimos meses.
    * Clientes que abandonan de manera repentina
    * Clientes que a pesar de mostrar "actitud NO churner" abandonan
- No churners:
    * Clientes que no hicieron churn mantienen o aumentan su actividad.
    * Pausas (Ej.: vacaciones, simplemente no tuvieron necesidad, etc.)

### 3. Interacciones con la app
- Churners:
   * Clientes con menor login, menor tiempo en app, menos clicks en productos financieros.
- No churners:
    * Clientes con mayor consistencia en el uso.

## <font color="#ff743d">**Churn**</font>

In [14]:
df_churn_man = df_churn.copy()

### <font color="#ffbf80">**Fechas**</font>

In [15]:
periods = 730

In [16]:
end_date = pd.to_datetime('2025-12-31')
end_date

Timestamp('2025-12-31 00:00:00')

In [17]:
start_date = end_date - pd.Timedelta(days=periods-1)
start_date

Timestamp('2024-01-02 00:00:00')

In [18]:
# Rango de fechas de acuerdo a la cantida de steps dentro del df_paysim
fechas = pd.date_range(start=start_date, periods=periods, freq="D")
fechas

DatetimeIndex(['2024-01-02', '2024-01-03', '2024-01-04', '2024-01-05',
               '2024-01-06', '2024-01-07', '2024-01-08', '2024-01-09',
               '2024-01-10', '2024-01-11',
               ...
               '2025-12-22', '2025-12-23', '2025-12-24', '2025-12-25',
               '2025-12-26', '2025-12-27', '2025-12-28', '2025-12-29',
               '2025-12-30', '2025-12-31'],
              dtype='datetime64[ns]', length=730, freq='D')

In [19]:
enum = [i for i in range(1, 731)]

In [20]:
df_fechas = pd.DataFrame({'step': enum, 'fecha': fechas})
df_fechas

,step,fecha
0,1,2024-01-02
1,2,2024-01-03
2,3,2024-01-04
3,4,2024-01-05
4,5,2024-01-06
...,...,...
725,726,2025-12-27
726,727,2025-12-28
727,728,2025-12-29
728,729,2025-12-30


In [21]:
df_fechas.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 730 entries, 0 to 729
Data columns (total 2 columns):
 #   Column  Non-Null Count  Dtype         
---  ------  --------------  -----         
 0   step    730 non-null    int64         
 1   fecha   730 non-null    datetime64[ns]
dtypes: datetime64[ns](1), int64(1)
memory usage: 11.5 KB


In [22]:
fechas_abandono = fechas[int((len(fechas) * 0.84)):]
fechas_abandono

DatetimeIndex(['2025-09-06', '2025-09-07', '2025-09-08', '2025-09-09',
               '2025-09-10', '2025-09-11', '2025-09-12', '2025-09-13',
               '2025-09-14', '2025-09-15',
               ...
               '2025-12-22', '2025-12-23', '2025-12-24', '2025-12-25',
               '2025-12-26', '2025-12-27', '2025-12-28', '2025-12-29',
               '2025-12-30', '2025-12-31'],
              dtype='datetime64[ns]', length=117, freq='D')

In [23]:
import random

np.random.seed(RANDOM_SEED)
random.seed(RANDOM_SEED) 

abandono_cust = []

for exited in df_churn_man['Exited']:
    if exited == 1:
        if random.random() < 0.95:
            # 95% de churners en los últimos 4 meses
            rand_idx = int(np.random.triangular(0, len(fechas_abandono)-1, len(fechas_abandono)-1))
            abandono_cust.append(fechas_abandono[rand_idx])
        else:
            # 5 % churners en cualquier fecha del período
            abandono_cust.append(random.choice(fechas))
    else:
        abandono_cust.append(np.nan)

In [24]:
df_churn_man['ExitDate'] = abandono_cust

In [25]:
df_churn_man

,RowNumber,CustomerId,Surname,CreditScore,Geography,Gender,Age,Tenure,Balance,NumOfProducts,HasCrCard,IsActiveMember,EstimatedSalary,Exited,ExitDate
0,1,15634602,Hargrave,619,France,Female,42,2,0.00,1,1,1,101348.88,1,2025-11-15
1,2,15647311,Hill,608,Spain,Female,41,1,83807.86,1,0,1,112542.58,0,NaT
2,3,15619304,Onio,502,France,Female,42,8,159660.80,3,1,0,113931.57,1,2025-12-28
3,4,15701354,Boni,699,France,Female,39,1,0.00,2,0,0,93826.63,0,NaT
4,5,15737888,Mitchell,850,Spain,Female,43,2,125510.82,1,1,1,79084.10,0,NaT
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
9995,9996,15606229,Obijiaku,771,France,Male,39,5,0.00,2,1,0,96270.64,0,NaT
9996,9997,15569892,Johnstone,516,France,Male,35,10,57369.61,1,1,1,101699.77,0,NaT
9997,9998,15584532,Liu,709,France,Female,36,7,0.00,1,0,1,42085.58,1,2025-12-18
9998,9999,15682355,Sabbatini,772,Germany,Male,42,3,75075.31,2,1,0,92888.52,1,2025-11-28


### <font color="#ffbf80">**Clusteres iniciales**</font>

Se asignarán scores de acuerdo a distintos aspectos del cliente:

* Tiempo: Edad y Antigüedad
* Riqueza: Balance, Salario Estimado y Puntaje Crediticio
* Compromiso: Número de productos, si Cuenta con Tarjeta de Crédito y si es Miembro Activo

Esto permitirá asignar un puntaje a cada cliente referente a cada aspecto, esto se utilizará luego para realizar una clusterización, y usar los clusters para asignar posteriormente las transacciones e interacciones con la app.

In [26]:
df_clusters = df_churn_man.copy()

In [27]:
df_clusters['cluster_time'] = pd.qcut(df_clusters['Tenure']*0.6 + df_clusters['Age']*0.4, q=3, labels=['Low','Medium','High'])

In [28]:
df_clusters['wealth_score'] = df_clusters['Balance']*0.5 + df_clusters['EstimatedSalary']*0.4 + df_clusters['CreditScore']*0.1

df_clusters['cluster_wealth'] = pd.qcut(df_clusters['wealth_score'], q=3, labels=['Low','Medium','High'])

In [29]:
df_clusters['engagement_score'] = df_clusters['NumOfProducts']*0.5 + df_clusters['HasCrCard']*0.2 + df_clusters['IsActiveMember']*0.3
df_clusters['cluster_engagement'] = pd.qcut(df_clusters['engagement_score'], q=3, labels=['Low','Medium','High'])

In [30]:
df_clusters

,RowNumber,CustomerId,Surname,CreditScore,Geography,Gender,Age,Tenure,Balance,NumOfProducts,HasCrCard,IsActiveMember,EstimatedSalary,Exited,ExitDate,cluster_time,wealth_score,cluster_wealth,engagement_score,cluster_engagement
0,1,15634602,Hargrave,619,France,Female,42,2,0.00,1,1,1,101348.88,1,2025-11-15,Medium,40601.452,Low,1.0,Low
1,2,15647311,Hill,608,Spain,Female,41,1,83807.86,1,0,1,112542.58,0,NaT,Medium,86981.762,Medium,0.8,Low
2,3,15619304,Onio,502,France,Female,42,8,159660.80,3,1,0,113931.57,1,2025-12-28,High,125453.228,High,1.7,High
3,4,15701354,Boni,699,France,Female,39,1,0.00,2,0,0,93826.63,0,NaT,Low,37600.552,Low,1.0,Low
4,5,15737888,Mitchell,850,Spain,Female,43,2,125510.82,1,1,1,79084.10,0,NaT,Medium,94474.050,Medium,1.0,Low
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
9995,9996,15606229,Obijiaku,771,France,Male,39,5,0.00,2,1,0,96270.64,0,NaT,Medium,38585.356,Low,1.2,Medium
9996,9997,15569892,Johnstone,516,France,Male,35,10,57369.61,1,1,1,101699.77,0,NaT,High,69416.313,Medium,1.0,Low
9997,9998,15584532,Liu,709,France,Female,36,7,0.00,1,0,1,42085.58,1,2025-12-18,Medium,16905.132,Low,0.8,Low
9998,9999,15682355,Sabbatini,772,Germany,Male,42,3,75075.31,2,1,0,92888.52,1,2025-11-28,Medium,74770.263,Medium,1.2,Medium


In [31]:
import pickle

In [32]:
factores = {
    'Tenure': 0.6,
    'Age': 0.4,
    'Balance': 0.5,
    'EstimatedSalary': 0.4,
    'CreditScore': 0.1,
    'NumOfProducts': 0.5,
    'HasCrCard': 0.2,
    'IsActiveMember': 0.3,
    'ClusterWealth': 0.4,
    'ClusterEngagement': 0.4,
    'ClusterTime': 0.2
}

In [33]:
with open(ARTIFACTS_PATH / 'factores_clusters.pkl', 'wb') as f:
    pickle.dump(factores, f)

In [34]:
# Normalizar los scores para que estén en la misma escala
from sklearn.preprocessing import MinMaxScaler

scaler = MinMaxScaler()

df_clusters[['wealth_score_norm', 'engagement_score_norm']] = scaler.fit_transform(df_clusters[['wealth_score', 'engagement_score']])

time_mapping = {'Low': 0, 'Medium': 0.5, 'High': 1}
df_clusters['cluster_time_num'] = df_clusters['cluster_time'].map(time_mapping).astype(float)

df_clusters.head()

,RowNumber,CustomerId,Surname,CreditScore,Geography,Gender,Age,Tenure,Balance,NumOfProducts,HasCrCard,IsActiveMember,EstimatedSalary,Exited,ExitDate,cluster_time,wealth_score,cluster_wealth,engagement_score,cluster_engagement,wealth_score_norm,engagement_score_norm,cluster_time_num
0,1,15634602,Hargrave,619,France,Female,42,2,0.00,1,1,1,101348.88,1,2025-11-15,Medium,40601.452,Low,1.0,Low,0.220009,0.25,0.5
1,2,15647311,Hill,608,Spain,Female,41,1,83807.86,1,0,1,112542.58,0,NaT,Medium,86981.762,Medium,0.8,Low,0.471943,0.15,0.5
2,3,15619304,Onio,502,France,Female,42,8,159660.80,3,1,0,113931.57,1,2025-12-28,High,125453.228,High,1.7,High,0.680918,0.60,1.0
3,4,15701354,Boni,699,France,Female,39,1,0.00,2,0,0,93826.63,0,NaT,Low,37600.552,Low,1.0,Low,0.203708,0.25,0.0
4,5,15737888,Mitchell,850,Spain,Female,43,2,125510.82,1,1,1,79084.10,0,NaT,Medium,94474.050,Medium,1.0,Low,0.512641,0.25,0.5


In [35]:
with open(ARTIFACTS_PATH / 'scaler_clusters.pkl', 'wb') as f:
    pickle.dump(scaler, f)

In [36]:
df_clusters['final_score'] = 0.4 * df_clusters['wealth_score_norm'] + 0.4 * df_clusters['engagement_score_norm'] + 0.2 * df_clusters['cluster_time_num']

In [37]:
df_clusters.describe()

,RowNumber,CustomerId,CreditScore,Age,Tenure,Balance,NumOfProducts,HasCrCard,IsActiveMember,EstimatedSalary,Exited,ExitDate,wealth_score,engagement_score,wealth_score_norm,engagement_score_norm,cluster_time_num,final_score
count,10000.00000,1.000000e+04,10000.000000,10000.000000,10000.000000,10000.000000,10000.000000,10000.00000,10000.000000,10000.000000,10000.000000,2037,10000.000000,10000.000000,10000.000000,10000.000000,10000.000000,10000.000000
mean,5000.50000,1.569094e+07,650.528800,38.921800,5.012800,76485.889288,1.530200,0.70550,0.515100,100090.239881,0.203700,2025-11-07 03:58:13.961708544,78344.093476,1.060730,0.425024,0.280365,0.495050,0.381166
min,1.00000,1.556570e+07,350.000000,18.000000,0.000000,0.000000,1.000000,0.00000,0.000000,11.580000,0.000000,2024-01-28 00:00:00,98.528000,0.500000,0.000000,0.000000,0.000000,0.006705
25%,2500.75000,1.562853e+07,584.000000,32.000000,3.000000,0.000000,1.000000,0.00000,0.000000,51002.110000,0.000000,2025-10-28 00:00:00,50439.055000,0.700000,0.273446,0.100000,0.000000,0.290167
50%,5000.50000,1.569074e+07,652.000000,37.000000,5.000000,97198.540000,1.000000,1.00000,1.000000,100193.915000,0.000000,2025-11-26 00:00:00,78047.218000,1.000000,0.423412,0.250000,0.500000,0.377777
75%,7500.25000,1.575323e+07,718.000000,44.000000,7.000000,127644.240000,2.000000,1.00000,1.000000,149388.247500,0.000000,2025-12-14 00:00:00,108695.399500,1.300000,0.589890,0.400000,1.000000,0.465492
max,10000.00000,1.581569e+07,850.000000,92.000000,10.000000,250898.090000,4.000000,1.00000,1.000000,199992.480000,1.000000,2025-12-30 00:00:00,184195.248000,2.500000,1.000000,1.000000,1.000000,0.912002
std,2886.89568,7.193619e+04,96.653299,10.487806,2.892174,62397.405202,0.581654,0.45584,0.499797,57510.492818,0.402769,NaN,38999.037709,0.340671,0.211840,0.170335,0.406838,0.126725


### <font color="#ffbf80">**Clusteres K-Means**</font>

In [38]:
from sklearn.cluster import KMeans

kmeans = KMeans(n_clusters=5, random_state=42)
df_clusters['cluster_final'] = kmeans.fit_predict(df_clusters[['final_score']])

In [39]:
df_clusters

,RowNumber,CustomerId,Surname,CreditScore,Geography,Gender,Age,Tenure,Balance,NumOfProducts,HasCrCard,IsActiveMember,EstimatedSalary,Exited,ExitDate,cluster_time,wealth_score,cluster_wealth,engagement_score,cluster_engagement,wealth_score_norm,engagement_score_norm,cluster_time_num,final_score,cluster_final
0,1,15634602,Hargrave,619,France,Female,42,2,0.00,1,1,1,101348.88,1,2025-11-15,Medium,40601.452,Low,1.0,Low,0.220009,0.25,0.5,0.288004,2
1,2,15647311,Hill,608,Spain,Female,41,1,83807.86,1,0,1,112542.58,0,NaT,Medium,86981.762,Medium,0.8,Low,0.471943,0.15,0.5,0.348777,4
2,3,15619304,Onio,502,France,Female,42,8,159660.80,3,1,0,113931.57,1,2025-12-28,High,125453.228,High,1.7,High,0.680918,0.60,1.0,0.712367,3
3,4,15701354,Boni,699,France,Female,39,1,0.00,2,0,0,93826.63,0,NaT,Low,37600.552,Low,1.0,Low,0.203708,0.25,0.0,0.181483,1
4,5,15737888,Mitchell,850,Spain,Female,43,2,125510.82,1,1,1,79084.10,0,NaT,Medium,94474.050,Medium,1.0,Low,0.512641,0.25,0.5,0.405056,4
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
9995,9996,15606229,Obijiaku,771,France,Male,39,5,0.00,2,1,0,96270.64,0,NaT,Medium,38585.356,Low,1.2,Medium,0.209058,0.35,0.5,0.323623,2
9996,9997,15569892,Johnstone,516,France,Male,35,10,57369.61,1,1,1,101699.77,0,NaT,High,69416.313,Medium,1.0,Low,0.376529,0.25,1.0,0.450612,0
9997,9998,15584532,Liu,709,France,Female,36,7,0.00,1,0,1,42085.58,1,2025-12-18,Medium,16905.132,Low,0.8,Low,0.091292,0.15,0.5,0.196517,1
9998,9999,15682355,Sabbatini,772,Germany,Male,42,3,75075.31,2,1,0,92888.52,1,2025-11-28,Medium,74770.263,Medium,1.2,Medium,0.405611,0.35,0.5,0.402245,4


In [40]:
with open(ARTIFACTS_PATH / 'kmeans_clusters.pkl', 'wb') as f:
    pickle.dump(kmeans, f)

### <font color="#ffbf80">**Interpretación clusters K-Means**</font>

In [41]:
clusters_count = df_clusters['cluster_final'].value_counts()
clusters_count

cluster_final
4    2874
2    2646
0    2230
1    1434
3     816
Name: count, dtype: int64

In [42]:
df_clusters_count = pd.DataFrame({'index':clusters_count.index, 'count': clusters_count})

In [43]:
df_clusters_count

,index,count
cluster_final,,
4,4,2874
2,2,2646
0,0,2230
1,1,1434
3,3,816


In [44]:
centroids_clusters_kmeans = kmeans.cluster_centers_
df_centroides = pd.DataFrame(centroids_clusters_kmeans, columns=['final_score'])
df_centroides.reset_index(inplace=True)

In [45]:
df_centroides

,index,final_score
0,0,0.494466
1,1,0.191013
2,2,0.299533
3,3,0.624389
4,4,0.395506


In [46]:
df_centroides = df_centroides.merge(df_clusters_count, on='index')
df_centroides = df_centroides.rename(columns={'index': 'cluster'})

In [47]:
df_centroides.sort_values('final_score')

,cluster,final_score,count
1,1,0.191013,1434
2,2,0.299533,2646
4,4,0.395506,2874
0,0,0.494466,2230
3,3,0.624389,816


In [48]:
def proporcion_labels(df):
    """
    Esta función tomará un df que debe corresponder a un unico cluster y
    retornará las proporciones de cada etiqueta de los clusters utilizados inicialmente
    """
    
    cluster_labels = ['cluster_time', 'cluster_wealth', 'cluster_engagement']
    counts = {}

    for cluster in cluster_labels:
        counts[cluster] = df[cluster].value_counts(normalize=True)

    return pd.DataFrame(counts)

#### <font color='#FDD2A9'>**Cluster 0**</font>

In [49]:
cluster_0 = df_clusters[df_clusters['cluster_final'] == 0]
prop_cluster_0 = proporcion_labels(cluster_0)
prop_cluster_0

,cluster_time,cluster_wealth,cluster_engagement
Low,0.070852,0.123318,0.503139
Medium,0.311211,0.343049,0.134978
High,0.617937,0.533632,0.361883


#### <font color='#FDD2A9'>**Observaciones Cluster 0:**</font>

Los clientes del **Cluster 0** cuentan con las siguientes características:
* Cuenta con un **score general** de: **0.494466** (2do más alto en el ranking)
* **Cantidad de clientes** pertenencientes al cluster: **2230 clientes**
* *cluster_time:* 60% para clientes con un alto score de tiempo (edad y permanencia) y 30% para término medio.
* *cluster_wealth:* 50% concentrado en clientes con buena salud económica, 35% para término medio.
* *cluster_engagement:* 50% con poco engagement, 35% con alto engagement.

Este es un buen perfil de cliente, ya que son clientes antiguos y con buen poder adquisitivo

#### <font color='#FDD2A9'>**Cluster 1**</font>

In [50]:
cluster_1 = df_clusters[df_clusters['cluster_final'] == 1]
prop_cluster_1 = proporcion_labels(cluster_1)
prop_cluster_1

,cluster_time,cluster_wealth,cluster_engagement
Low,0.82636,0.734310,0.722455
Medium,0.16318,0.243375,0.156206
High,0.01046,0.022315,0.121339


#### <font color='#FDD2A9'>**Observaciones Cluster 1:**</font>

Los clientes del **Cluster 1** cuentan con las siguientes características:

* Cuenta con un **score general** de: **0.191013** (ranking más bajo)
* **Cantidad de clientes** pertenencientes al cluster: **1434 clientes**
* *cluster_time:* 82% para clientes con un bajo score de tiempo (edad y permanencia) y 16% para término medio.
* *cluster_wealth:* 73% concentrado en clientes con baja salud económica, 24% para término medio.
* *cluster_engagement:* 72% con poco engagement, 12% con engagement término medio y 12% con alto engagement.

Este es el perfi de cliente con menor valor para la empresa, son clientes nuevos, de bajo poder adquisitivo, y con poco compromiso

#### <font color='#FDD2A9'>**Cluster 2**</font>

In [51]:
cluster_2 = df_clusters[df_clusters['cluster_final'] == 2]
prop_cluster_2 = proporcion_labels(cluster_2)
prop_cluster_2

,cluster_time,cluster_wealth,cluster_engagement
Low,0.522298,0.432729,0.642857
Medium,0.387755,0.365835,0.167800
High,0.089947,0.201436,0.189342


#### <font color='#FDD2A9'>**Observaciones Cluster 2:**</font>

Los clientes del **Cluster 2** cuentan con las siguientes características:

* Cuenta con un **score general** de: **0.299533**
* **Cantidad de clientes** pertenencientes al cluster: **2646 clientes**
* *cluster_time:* 52% para clientes con un bajo score de tiempo (edad y permanencia) y 38% para término medio.
* *cluster_wealth:* 20% concentrado en clientes con buena salud económica, 36% para término medio, y 43% para baja riqueza.
* *cluster_engagement:* 64% con poco engagement, 16% con engagement término medio y 19% con alto engagement.

Este es el perfi de cliente tiene poco compromiso, pero con potencial para fidelizar, ya que la riqueza esta distribuida de forma relativamente pareja, y son clientes nuevos y/o jovenes.

#### <font color='#FDD2A9'>**Cluster 3**</font>

In [52]:
cluster_3 = df_clusters[df_clusters['cluster_final'] == 3]
prop_cluster_3 = proporcion_labels(cluster_3)
prop_cluster_3

,cluster_time,cluster_wealth,cluster_engagement
Low,0.009804,0.030637,0.215686
Medium,0.151961,0.197304,0.123775
High,0.838235,0.772059,0.660539


#### <font color='#FDD2A9'>**Observaciones Cluster 3:**</font>

Los clientes del **Cluster 3** cuentan con las siguientes características:
	
* Cuenta con un **score general** de: **0.624389** (ranking más alto)
* **Cantidad de clientes** pertenencientes al cluster: **816 clientes** (cluster más pequeño)
* *cluster_time:* 83% para clientes con un alto score de tiempo (edad y permanencia) y 15% para término medio.
* *cluster_wealth:* 77% concentrado en clientes con buena salud económica, 19% para término medio, y 3% para baja riqueza.
* *cluster_engagement:* 66% con alto engagement, 12% con engagement término medio y 21% con bajo engagement.

Este es el perfil de cliente **"VIP"**. Son clientes antiguos y con muy buena salud económica (posiblemente también sean los clientes más adultos). Y con buen nivel de compromiso.

#### <font color='#FDD2A9'>**Cluster 4**</font>

In [53]:
cluster_4 = df_clusters[df_clusters['cluster_final'] == 4]
prop_cluster_4 = proporcion_labels(cluster_4)
prop_cluster_4

,cluster_time,cluster_wealth,cluster_engagement
Low,0.218163,0.290536,0.585247
Medium,0.452679,0.379610,0.153445
High,0.329158,0.329854,0.261308


#### <font color='#FDD2A9'>**Observaciones Cluster 4:**</font>

Los clientes del **Cluster 4** cuentan con las siguientes características:
	
* Cuenta con un **score general** de: **0.395506**
* **Cantidad de clientes** pertenencientes al cluster: **2874 clientes** (cluster más grande)
* *cluster_time:* 32% para clientes con un alto score de tiempo (edad y permanencia) y 45% para término medio y 21% de bajo score de tiempo.
* *cluster_wealth:* 32% de clientes con buena salud económica, 37% para término medio, y 29% para baja riqueza.
* *cluster_engagement:* 26% con alto engagement, 15% con engagement término medio y 58% con bajo engagement.

Este es el perfil de cliente **"Standard"**. Son clientes ni muy antiguos ni muy nuevos, ni muy ricos ni muy pobres, predomina un compromiso bajo.

In [54]:
etiquetas = {0: 'Valioso - Bajo compromiso',
             1: 'Poco Valor',
             2: 'Cliente potencial',
             3: 'VIP',
             4: 'Standard'}

In [55]:
with open(ARTIFACTS_PATH / 'labels_clusters.pkl', 'wb') as f:
    pickle.dump(etiquetas, f)

In [56]:
df_clusters['CustomerSegment'] = df_clusters['cluster_final'].map(etiquetas)

In [57]:
df_clusters.sample(20)

,RowNumber,CustomerId,Surname,CreditScore,Geography,Gender,Age,Tenure,Balance,NumOfProducts,HasCrCard,IsActiveMember,EstimatedSalary,Exited,ExitDate,cluster_time,wealth_score,cluster_wealth,engagement_score,cluster_engagement,wealth_score_norm,engagement_score_norm,cluster_time_num,final_score,cluster_final,CustomerSegment
6955,6956,15697042,Genovesi,738,Spain,Male,35,8,127290.61,1,1,0,16081.62,0,NaT,Medium,70151.753,Medium,0.7,Low,0.380524,0.10,0.5,0.292210,2,Cliente potencial
1626,1627,15588248,Hs?,617,France,Female,28,0,0.00,2,1,1,7597.83,1,2025-11-02,Low,3100.832,Low,1.5,High,0.016308,0.50,0.0,0.206523,1,Poco Valor
4400,4401,15567063,Manna,766,Germany,Female,34,6,106434.94,1,0,1,137995.66,1,2025-12-22,Medium,108492.334,High,0.8,Low,0.588787,0.15,0.5,0.395515,4,Standard
5713,5714,15643439,Ferguson,537,France,Male,47,10,0.00,2,0,1,25482.62,0,NaT,High,10246.748,Low,1.3,High,0.055124,0.40,1.0,0.382050,4,Standard
6997,6998,15689061,Davey,611,France,Male,68,5,82547.11,2,1,1,146448.01,0,NaT,High,99913.859,High,1.5,High,0.542190,0.50,1.0,0.616876,3,VIP
741,742,15716835,Rossi,546,France,Male,24,8,156325.38,1,1,1,125381.02,0,NaT,Low,128369.698,High,1.0,Low,0.696760,0.25,0.0,0.378704,4,Standard
2935,2936,15689425,Olejuru,687,Spain,Male,35,8,100988.39,2,1,0,22247.27,0,NaT,Medium,59461.803,Low,1.2,Medium,0.322457,0.35,0.5,0.368983,4,Standard
1665,1666,15775082,Stewart,749,France,Male,42,1,129776.72,2,0,1,143538.51,0,NaT,Medium,122378.664,High,1.3,High,0.664217,0.40,0.5,0.525687,0,Valioso - Bajo compromiso
3079,3080,15774727,Monaldo,757,Germany,Female,34,1,129398.01,2,0,0,44965.44,0,NaT,Low,82760.881,Medium,1.0,Low,0.449016,0.25,0.0,0.279606,2,Cliente potencial
1217,1218,15696900,Burns,505,Germany,Male,29,3,145541.56,2,1,1,58019.95,0,NaT,Low,96029.260,Medium,1.5,High,0.521089,0.50,0.0,0.408436,4,Standard


In [58]:
df_clients = df_clusters.copy()

In [59]:
df_clients.groupby('CustomerSegment')['CustomerSegment'].value_counts()

CustomerSegment
Cliente potencial            2646
Poco Valor                   1434
Standard                     2874
VIP                           816
Valioso - Bajo compromiso    2230
Name: count, dtype: int64

In [60]:
df_clients['CustomerSegment'].value_counts(normalize=True)

CustomerSegment
Standard                     0.2874
Cliente potencial            0.2646
Valioso - Bajo compromiso    0.2230
Poco Valor                   0.1434
VIP                          0.0816
Name: proportion, dtype: float64

In [61]:
churn_yes = df_clients[df_clients['Exited'] == 1]
churn_yes.groupby('CustomerSegment')['CustomerSegment'].value_counts()

CustomerSegment
Cliente potencial            380
Poco Valor                   178
Standard                     590
VIP                          329
Valioso - Bajo compromiso    560
Name: count, dtype: int64

### <font color="#ffbf80">**Generación de transacciones**</font>

A continuación, utilizando las etiquetas de ambos datasets, se asignarán las transacciones de manera aleatoria, pero respetando ciertas proporciones que reflejen la realidad.

**Nota:**

Las siguients proporciones fueron asignadas a mano, buscando una distribución lógica entre los distintos tipos de clientes y transacciones y como debería ser el comportamiento de estos.

Además, se asignarán factores aleatorios para introducir ruido estadístico al conjunto de datos

#### **Las siguients tablas se utilizarán como patrones a seguir en la asignación de transacciones a los clientes**

**Guía**

| Etiqueta	                  |	Número de transacciones	 |
|-----------------------------|--------------------------|
| VIP	                      | Alta	                 |
| Valioso - Bajo compromiso   | Media	                 |
| Standard              	  | Media	                 |
| Clientes potenciales	      | Baja	                 |
| Nuevos - Poco Valor         | Baja	                 |

**Transacciones**

| Tran Label  | Proporción |
|-------------|------------|
| Low         | 0.33       |
| High        | 0.30       |
| Medium      | 0.27       |
| VIP         | 0.10       |

**Distribución Transacciones**

| Cliente Label \ Tran Label | Low           | Medium          | High         | Vip        | Total      |
|----------------------------|---------------|-----------------|--------------|------------|------------|
| Nuevos - Poco Valor        | 10%           | 3%              | 1%           | 0.5%       | 14%        |
| Clientes potenciales       | 8%            | 8%              | 3%           | 1%         | 20%        |
| Standard              	 | 6.5%          | 5%              | 7%           | 1%         | 19.5%      |
| Valioso - Bajo compromiso  | 7.5%          | 8%              | 4%           | 1%         | 20.5%      |
| VIP	                     | 1%            | 3%              | 15%          | 6.5%       | 26%        |
| Total                      | 33%           | 27%             | 30%          | 10%        | 100%       |

# <font color="#6e8ebc">**Transactions - Generación de datos**</font>

In [62]:
import numpy as np
import pandas as pd

# --- CONFIGURACIÓN Y SEMILLA ---
RANDOM_SEED = 42
RNG = np.random.default_rng(RANDOM_SEED)

# --- HIPERPARÁMETROS GENERALES ---
TX_VOLUME_NOISE = 0.15   

# --- HIPERPARÁMETROS DE COMPORTAMIENTO ---

# 1. COMPORTAMIENTO DE DRENAJE (Signal - Churners)
PROB_CHURNER_DRAINS = 0.60        # 60% de los que se van, sacan su dinero antes.
DRAIN_WINDOW_DAYS = 60            # Ventana de pánico.
DRAIN_MULTIPLIER_MAX = 3.0        # Intensidad del pánico.

# 2. RUIDO DE GASTO (Noise - Non-Churners)
PROB_LOYAL_BIG_SPENDER = 0.15     # 15% de fieles hacen gastos grandes esporádicos.
LOYAL_SPIKE_MULTIPLIER = 2.5      

# 3. (NUEVO) RUIDO DE INACTIVIDAD/PAUSAS (Noise - Non-Churners que parecen Churners)
PROB_PAUSE_LOYAL = 0.15           # 15% de los fieles tendrán una "laguna" de inactividad.
PAUSE_MIN_MONTHS = 1              # Duración mínima de la pausa.
PAUSE_MAX_MONTHS = 3              # Duración máxima (vacaciones largas, temporada baja).

# Tipos de transacción
TX_TYPES = ['PAYMENT', 'TRANSFER', 'CASH_OUT', 'DEBIT', 'CASH_IN']

# Multiplicadores
TX_TYPE_MULTIPLIERS = {
    'PAYMENT': 0.3,    
    'TRANSFER': 1.0,   
    'CASH_OUT': 0.5,   
    'DEBIT': 0.15,     
    'CASH_IN': 1.0     
}

# --- PARÁMETROS POR CLUSTER ---
tx_cluster_params = {
    "Poco Valor": {
        "base_tx_per_month": 3,
        "salary_ratio": 0.12,     
        "amount_volatility": 0.2, 
        "probs_type": [0.40, 0.10, 0.10, 0.30, 0.10] 
    },
    "Cliente potencial": {
        "base_tx_per_month": 6,
        "salary_ratio": 0.25,  
        "amount_volatility": 0.45,
        "probs_type": [0.35, 0.20, 0.10, 0.25, 0.10]
    },
    "Standard": {
        "base_tx_per_month": 10,
        "salary_ratio": 0.15,  
        "amount_volatility": 0.25, 
        "probs_type": [0.30, 0.30, 0.10, 0.20, 0.10]
    },
    "Valioso - Bajo compromiso": {
        "base_tx_per_month": 8,
        "salary_ratio": 0.20, 
        "amount_volatility": 0.30,
        "probs_type": [0.20, 0.40, 0.20, 0.10, 0.10]
    },
    "VIP": {
        "base_tx_per_month": 20,
        "salary_ratio": 0.4, 
        "amount_volatility": 0.65, 
        "probs_type": [0.15, 0.50, 0.15, 0.05, 0.15]
    }
}

# --- PARÁMETROS DE VOLUMEN (DECAY) ---
LAST_N_MONTHS_DECAY = 4
DECAY_FINAL_FRAC_TX = 0.10
PROB_CHURN_REPENTINO_TX = 0.20

# --- PREPARACIÓN DEL GRID ---
# (Asumiendo que start_date y end_date están definidos en tu entorno)
freq = "M"
months = pd.period_range(start=start_date, end=end_date, freq=freq).to_timestamp('M')

# Cargar clientes
clients_tx = df_clients[['CustomerId', 'CustomerSegment', 'EstimatedSalary', 'Exited', 'ExitDate', 'Tenure']].copy()
clients_tx['Tenure'] = clients_tx['Tenure'].astype(float)
mean_salary = clients_tx['EstimatedSalary'].mean()
clients_tx['EstimatedSalary'] = clients_tx['EstimatedSalary'].fillna(mean_salary)

# Grid Cliente x Mes
cust_ids = clients_tx['CustomerId'].values
n_cust = len(cust_ids)
n_months = len(months)

grid_tx = pd.DataFrame({
    'CustomerId': np.repeat(cust_ids, n_months),
    'month': np.tile(months.to_numpy(), n_cust)
})
grid_tx = grid_tx.merge(clients_tx, on='CustomerId', how='left')

# Mapa de churners repentinos
churners_repentinos_tx = (RNG.random(len(clients_tx[clients_tx['Exited'] == 1])) < PROB_CHURN_REPENTINO_TX)
client_repentino_map_tx = clients_tx[clients_tx['Exited'] == 1].set_index('CustomerId').assign(is_repentino=churners_repentinos_tx)['is_repentino'].to_dict()

# ==============================================================================
# === (NUEVO) GENERACIÓN DE PAUSAS PARA NO-CHURNERS ===
# ==============================================================================
# Seleccionamos clientes fieles para aplicarles "silencios" temporales
non_churners_ids = clients_tx[clients_tx['Exited'] == 0]['CustomerId'].values
ids_with_pause = RNG.choice(non_churners_ids, size=int(len(non_churners_ids) * PROB_PAUSE_LOYAL), replace=False)

pause_rows = []
total_months_idx = len(months)

for cid in ids_with_pause:
    # Duración aleatoria
    duration = RNG.integers(PAUSE_MIN_MONTHS, PAUSE_MAX_MONTHS + 1)
    
    # Elegir mes de inicio (asegurando que entre en el rango)
    if total_months_idx > duration:
        start_idx = RNG.integers(0, total_months_idx - duration)
        pause_start = months[start_idx]
        pause_end = months[start_idx + duration - 1]
        pause_rows.append({'CustomerId': cid, 'pause_start': pause_start, 'pause_end': pause_end})

# Mergeamos esta info al grid principal
if pause_rows:
    df_pauses = pd.DataFrame(pause_rows)
    grid_tx = grid_tx.merge(df_pauses, on='CustomerId', how='left')
else:
    grid_tx['pause_start'] = pd.NaT
    grid_tx['pause_end'] = pd.NaT

# ==============================================================================

# --- 1. CÁLCULO DE VOLUMEN (LAMBDA) CON PAUSAS ---
def compute_tx_lambda(row):
    params = tx_cluster_params.get(row['CustomerSegment'], tx_cluster_params['Standard'])
    base_lam = params['base_tx_per_month']
    
    # Factor Antigüedad
    tenure_factor = (row['Tenure'] / 60.0) * 0.2 
    mean_lam = base_lam * (1 + tenure_factor)
    
    # Ruido Mensual
    noise = np.clip(RNG.normal(loc=1.0, scale=TX_VOLUME_NOISE), 0.7, 1.3)
    lam = max(0.1, mean_lam * noise)
    
    # A) LÓGICA DE PAUSAS (Non-Churners) - NUEVO
    if row['Exited'] == 0:
        if pd.notna(row['pause_start']):
            # Si el mes actual cae en su pausa, lambda = 0 (Silencio)
            if row['pause_start'] <= row['month'] <= row['pause_end']:
                return 0.0
    
    # B) LÓGICA DE DECAY (Churners)
    if row['Exited'] == 1:
        exit_date = pd.to_datetime(row['ExitDate'])
        month_end = row['month'] + pd.offsets.MonthEnd(0)
        
        if month_end >= exit_date:
            return 0.0
            
        is_repentino = client_repentino_map_tx.get(row['CustomerId'], False)
        
        if not is_repentino:
            diff = (exit_date.to_period('M') - row['month'].to_period('M')).n
            if diff <= LAST_N_MONTHS_DECAY and diff > 0:
                frac = diff / float(LAST_N_MONTHS_DECAY)
                decay_factor = (DECAY_FINAL_FRAC_TX + (1 - DECAY_FINAL_FRAC_TX) * frac)
                lam = lam * decay_factor
                
    return lam

print("Generando volumen de transacciones (con pausas simuladas)...")
grid_tx['lambda_tx'] = grid_tx.apply(compute_tx_lambda, axis=1)
grid_tx['n_tx'] = RNG.poisson(grid_tx['lambda_tx'].values)
grid_tx = grid_tx[grid_tx['n_tx'] > 0].reset_index(drop=True)

print("Expandiendo a nivel transacción...")
rep_counts = grid_tx['n_tx'].values
total_txs = rep_counts.sum()

tx_customer = np.repeat(grid_tx['CustomerId'].values, rep_counts)
tx_month = np.repeat(grid_tx['month'].values.astype('datetime64[ns]'), rep_counts)
tx_cluster_arr = np.repeat(grid_tx['CustomerSegment'].values, rep_counts)
tx_salary_arr = np.repeat(grid_tx['EstimatedSalary'].values, rep_counts)

# --- 2. ASIGNACIÓN INICIAL DE MONTOS ---
print("Asignando montos base...")

tx_types = np.empty(total_txs, dtype=object)
tx_amounts = np.zeros(total_txs, dtype=float)

for clust_name, params in tx_cluster_params.items():
    mask = (tx_cluster_arr == clust_name)
    n_cluster = mask.sum()
    if n_cluster == 0: continue
    
    # Tipos
    types_chosen = RNG.choice(TX_TYPES, size=n_cluster, p=params['probs_type'])
    tx_types[mask] = types_chosen
    
    # Montos
    salaries = tx_salary_arr[mask]
    monthly_income = salaries / 12.0
    
    base_target = monthly_income * params['salary_ratio']
    type_factors = np.vectorize(TX_TYPE_MULTIPLIERS.get)(types_chosen)
    target_amounts = base_target * type_factors
    
    volatility = params['amount_volatility']
    noise = RNG.normal(loc=1.0, scale=volatility, size=n_cluster)
    noise = np.clip(noise, 0.5, 2.0) 
    
    final_amounts = target_amounts * noise
    final_amounts = np.clip(final_amounts, 5.0, salaries / 3.0)
    tx_amounts[mask] = np.round(final_amounts, 2)

# Fechas base
month_starts = pd.to_datetime(tx_month).to_period('M').to_timestamp()
month_ends = (pd.to_datetime(tx_month).to_period('M') + 1).to_timestamp() - pd.Timedelta(days=1)
days_in_month = (month_ends - month_starts).days + 1
rand_days = RNG.integers(0, days_in_month, size=total_txs)
tx_dates = month_starts + pd.to_timedelta(rand_days, unit='D')

# --- 3. DATAFRAME TEMPORAL Y LÓGICA DE DRENAJE ---
df_synth = pd.DataFrame({
    'CustomerId': tx_customer,
    'TransactionDate': tx_dates,
    'Amount': tx_amounts,
    'TransactionType': tx_types,
    'EstimatedSalary': tx_salary_arr
})

df_synth = df_synth.merge(clients_tx[['CustomerId', 'Exited', 'ExitDate']], on='CustomerId', how='left')
df_synth['ExitDate'] = pd.to_datetime(df_synth['ExitDate'])

# ==============================================================================
# === LÓGICA DE DRENAJE (CASH OUT SPIKES) ===
# ==============================================================================
print("\n🔥 Aplicando lógica de 'Drenaje de Cuenta' y 'Gastos Extra'...")

# A) SEÑAL: CHURNERS QUE DRENAN
churner_ids = clients_tx[clients_tx['Exited'] == 1]['CustomerId'].values
draining_churners = RNG.choice(churner_ids, size=int(len(churner_ids) * PROB_CHURNER_DRAINS), replace=False)

df_synth['days_to_exit'] = (df_synth['ExitDate'] - df_synth['TransactionDate']).dt.days

mask_drain_signal = (
    (df_synth['CustomerId'].isin(draining_churners)) & 
    (df_synth['TransactionType'] == 'CASH_OUT') & 
    (df_synth['days_to_exit'] >= 0) & 
    (df_synth['days_to_exit'] <= DRAIN_WINDOW_DAYS)
)

factor_urgencia = 1 - (df_synth.loc[mask_drain_signal, 'days_to_exit'] / DRAIN_WINDOW_DAYS)
multipliers_signal = 1.0 + (factor_urgencia * (DRAIN_MULTIPLIER_MAX - 1.0))
df_synth.loc[mask_drain_signal, 'Amount'] *= multipliers_signal
print(f"   -> {mask_drain_signal.sum()} transacciones de Churners infladas (Señal).")


# B) RUIDO: CLIENTES FIELES QUE GASTAN MUCHO
loyal_ids = clients_tx[clients_tx['Exited'] == 0]['CustomerId'].values
big_spenders_loyal = RNG.choice(loyal_ids, size=int(len(loyal_ids) * PROB_LOYAL_BIG_SPENDER), replace=False)

mask_potential_noise = (
    (df_synth['CustomerId'].isin(big_spenders_loyal)) & 
    (df_synth['TransactionType'] == 'CASH_OUT')
)
noise_indices = df_synth[mask_potential_noise].sample(frac=0.10, random_state=RANDOM_SEED).index
df_synth.loc[noise_indices, 'Amount'] *= LOYAL_SPIKE_MULTIPLIER
print(f"   -> {len(noise_indices)} transacciones de Fieles infladas (Ruido).")

# ==============================================================================

# --- LIMPIEZA FINAL ---
df_synth['Amount'] = np.clip(df_synth['Amount'], 0, df_synth['EstimatedSalary'])
df_synth['Amount'] = df_synth['Amount'].round(2)
df_synth['TransactionId'] = np.arange(len(df_synth)).astype(str)

df_transactions_synth = df_synth[['TransactionId', 'CustomerId', 'TransactionDate', 'Amount', 'TransactionType']]
df_transactions_synth = df_transactions_synth.sort_values(['CustomerId', 'TransactionDate']).reset_index(drop=True)

print("\n--- Resumen Final Transacciones ---")
print(df_transactions_synth.groupby('TransactionType')['Amount'].describe().round(2))

Generando volumen de transacciones (con pausas simuladas)...
Expandiendo a nivel transacción...
Asignando montos base...

🔥 Aplicando lógica de 'Drenaje de Cuenta' y 'Gastos Extra'...
   -> 1240 transacciones de Churners infladas (Señal).
   -> 2996 transacciones de Fieles infladas (Ruido).

--- Resumen Final Transacciones ---
                    count     mean      std   min      25%      50%      75%  \
TransactionType                                                                
CASH_IN          212541.0  2510.04  2375.15  3.86   934.77  1830.67  3115.08   
CASH_OUT         254483.0  1237.42  1163.61  3.86   496.96   931.57  1531.98   
DEBIT            319007.0   258.27   233.63  3.86   101.68   207.57   341.59   
PAYMENT          513901.0   578.49   550.22  3.86   222.16   447.05   743.00   
TRANSFER         640024.0  2644.28  2420.70  3.86  1052.73  1957.28  3218.48   

                      max  
TransactionType            
CASH_IN          13328.61  
CASH_OUT         16660.78 

In [63]:
# 1. Calcular la fecha de la última transacción por cliente en df_transactions_synth
# La columna de fecha en df_transactions_synth es 'transaction_date'.
last_tx_date = df_transactions_synth.groupby('CustomerId')['TransactionDate'].max().reset_index(name='LastTransactionDate')

# 2. Filtrar clientes que abandonaron y unirlos con su última transacción
df_churned = df_clients[df_clients['Exited'] == 1].copy()
# Seleccionamos las columnas relevantes y realizamos el merge
df_churned_check = df_churned[['CustomerId', 'ExitDate', 'Exited']].merge(
    last_tx_date, on='CustomerId', how='left'
)

# 3. Realizar la comprobación de integridad
# El error ocurre si la ExitDate es ANTES de la LastTransactionDate.
df_churned_check['TemporalIntegrityError'] = df_churned_check['ExitDate'] < df_churned_check['LastTransactionDate']

# 4. Reportar los casos problemáticos (donde el error es True)
integrity_violations = df_churned_check[df_churned_check['TemporalIntegrityError'] == True]

print(f"Clientes que abandonaron ('Exited'=1): {df_churned.shape[0]}")
print(f"Clientes con violación de integridad temporal: {integrity_violations.shape[0]}")

if integrity_violations.empty:
    print("\n✅ ¡Validación completada! No se encontraron transacciones después de las fechas de salida registradas.")
else:
    print("\n🚨 ¡ATENCIÓN! Se encontraron violaciones de integridad. Primeros 5 casos:")
    # Usamos .dt.date para imprimir las fechas de forma limpia
    violations_clean = integrity_violations[['CustomerId', 'ExitDate', 'LastTransactionDate']].copy()
    violations_clean['ExitDate'] = violations_clean['ExitDate'].dt.date
    violations_clean['LastTransactionDate'] = violations_clean['LastTransactionDate'].dt.date
    
    print(violations_clean.head())
    print("\nEstas filas indican que el cliente tuvo actividad de transacción posterior a su fecha de salida ('ExitDate').")

Clientes que abandonaron ('Exited'=1): 2037
Clientes con violación de integridad temporal: 0

✅ ¡Validación completada! No se encontraron transacciones después de las fechas de salida registradas.


In [64]:
df_transactions_synth.sort_values(by='TransactionDate', inplace=True)

# <font color="#6e8ebc">**App Interaction - Generación de datos**</font>

In [65]:
import numpy as np
import pandas as pd
from datetime import datetime
import uuid

# --- REPRODUCIBILIDAD ---
# Es fundamental para que los resultados sean consistentes en cada ejecución.
RANDOM_SEED = 42
RNG = np.random.default_rng(RANDOM_SEED)

# --- CONFIGURACIÓN GLOBAL DE RUIDO ---
# Estos parámetros añaden variabilidad y realismo al comportamiento de los usuarios.
LAMBDA_NOISE_SIGMA = 0.20 # Variabilidad en la actividad mensual de cada cliente.
DECAY_NOISE_SIGMA = 0.25   # Variabilidad en el decaimiento de actividad de los churners.
PROB_NOISE_SIGMA = 0.15    # Variabilidad en las acciones que realiza un usuario en cada sesión.

# --- DEFINICIONES DE TIEMPO Y CLUSTERS ---
# (Asegúrate de que 'start_date' y 'end_date' estén definidas previamente)
freq = "M"  # Agruparemos la actividad por Meses.
months = pd.period_range(start=start_date, end=end_date, freq=freq).to_timestamp('M')

# nombres de clusters según tu df_clients
clusters = [
    "Poco Valor",
    "Cliente potencial",
    "Standard",
    "Valioso - Bajo compromiso",
    "VIP"
]

# parámetros base por cluster (ejemplo; ajustalos)
cluster_params = {
    "Poco Valor": {
        "base_sessions_per_month": 1.5,
        "duration_mu": 2.0,  # log-normal mu
        "duration_sigma": 0.6,
        "p_transfer": 0.02,
        "p_payment": 0.05,
        "p_invest": 0.01,
        "p_push_open": 0.10,
        "p_failed_login": 0.01,
    },
    "Cliente potencial": {
        "base_sessions_per_month": 3.0,
        "duration_mu": 2.3,
        "duration_sigma": 0.7,
        "p_transfer": 0.06,
        "p_payment": 0.08,
        "p_invest": 0.02,
        "p_push_open": 0.15,
        "p_failed_login": 0.02,
    },
    "Standard": {
        "base_sessions_per_month": 4.0,
        "duration_mu": 2.7,
        "duration_sigma": 0.8,
        "p_transfer": 0.10,
        "p_payment": 0.12,
        "p_invest": 0.03,
        "p_push_open": 0.18,
        "p_failed_login": 0.03,
    },
    "Valioso - Bajo compromiso": {
        "base_sessions_per_month": 3.5,
        "duration_mu": 2.6,
        "duration_sigma": 0.8,
        "p_transfer": 0.12,
        "p_payment": 0.10,
        "p_invest": 0.05,
        "p_push_open": 0.12,
        "p_failed_login": 0.02,
    },
    "VIP": {
        "base_sessions_per_month": 8.0,
        "duration_mu": 3.0,
        "duration_sigma": 0.9,
        "p_transfer": 0.25,
        "p_payment": 0.20,
        "p_invest": 0.15,
        "p_push_open": 0.35,
        "p_failed_login": 0.01,
    }
}


# --- PARÁMETROS DE COMPORTAMIENTO DE CHURN Y TENURE ---
prob_churn_repentino_app = 0.20 # 20% de los churners se irán "de golpe".
last_n_months_decay = 6       # Ventana de 6 meses donde la actividad de un churner decae.
decay_final_frac = 0.05       # La actividad cae hasta un 5% de lo normal justo antes de irse.
tenure_max = df_clients['Tenure'].max()
tenure_weight = 0.5           # Influencia de la antigüedad del cliente en su actividad.

# ==============================================================================
# ===            HIPERPARÁMETROS DE PAUSA PARA NO CHURNERS                   ===
# ==============================================================================
PROB_PAUSA_NO_CHURNER_APP = 0.15  # 15% de los clientes fieles tendrán una pausa.
PAUSA_DURACION_MIN_MESES = 2       # Duración mínima de la pausa en meses.
PAUSA_DURACION_MAX_MESES = 4       # Duración máxima de la pausa en meses.
# ==============================================================================

# --- 1) PREPARAR BASE CLIENTE x MES ---
clients = df_clients[['CustomerId','CustomerSegment','Exited','ExitDate','Tenure']].copy()
clients['Tenure'] = clients['Tenure'].astype(float)

# Crear tabla con una fila por cada cliente y cada mes de la simulación.
cust_ids = clients['CustomerId'].values
n_cust = len(cust_ids)
n_months = len(months)
grid = pd.DataFrame({
    'CustomerId': np.repeat(cust_ids, n_months),
    'month': np.tile(months.to_numpy(), n_cust)
})
grid = grid.merge(clients, on='CustomerId', how='left')

# Mapa de churners repentinos (sin cambios).
churners_repentinos = (RNG.random(len(clients[clients['Exited'] == 1])) < prob_churn_repentino_app)
client_repentino_map = clients[clients['Exited'] == 1].set_index('CustomerId').assign(is_repentino=churners_repentinos)['is_repentino'].to_dict()


# ==============================================================================
# === (NUEVO) PRE-CÁLCULO DE LAS PAUSAS DE ACTIVIDAD PARA CLIENTES FIELES ===
# ==============================================================================
# 1. Seleccionar qué clientes no-churners tendrán una pausa.
no_churners = clients[clients['Exited'] == 0].copy()
clientes_con_pausa_ids = no_churners.sample(frac=PROB_PAUSA_NO_CHURNER_APP, random_state=RANDOM_SEED)['CustomerId']

# 2. Para cada cliente seleccionado, definir el inicio y fin de su pausa.
pause_info = []
total_meses = len(months)
for cid in clientes_con_pausa_ids:
    duracion_pausa = np.random.randint(PAUSA_DURACION_MIN_MESES, PAUSA_DURACION_MAX_MESES + 1)
    if total_meses > duracion_pausa:
        start_offset = np.random.randint(0, total_meses - duracion_pausa)
        pausa_inicio = months[start_offset]
        pausa_fin = months[start_offset + duracion_pausa - 1]
        pause_info.append({'CustomerId': cid, 'pause_start': pausa_inicio, 'pause_end': pausa_fin})

# 3. Unir la información de las pausas a la tabla principal `grid`.
if pause_info:
    df_pausas = pd.DataFrame(pause_info)
    grid = grid.merge(df_pausas, on='CustomerId', how='left')
else:
    grid['pause_start'] = pd.NaT
    grid['pause_end'] = pd.NaT

grid['pause_start'] = grid['pause_start'].fillna(pd.NaT)
grid['pause_end'] = grid['pause_end'].fillna(pd.NaT)
# ==============================================================================


# --- FUNCIÓN DE CÁLCULO DE ACTIVIDAD MENSUAL (MODIFICADA) ---
def compute_lambda_row(row):
    params = cluster_params.get(row['CustomerSegment'], cluster_params['Standard'])
    
    # Cálculo de Lambda Base con ruido (sin cambios).
    base_lam = params['base_sessions_per_month']
    tenure_factor = (row['Tenure'] / tenure_max) * tenure_weight
    mean_lam = base_lam * (1 + tenure_factor)
    noise_factor_base = np.clip(RNG.normal(loc=1.0, scale=LAMBDA_NOISE_SIGMA), 0.5, 2.0)
    lam = max(0.01, mean_lam * noise_factor_base)

    # ==========================================================================
    # === (MODIFICADO) LÓGICA PARA NO-CHURNERS CON PAUSAS ===
    # ==========================================================================
    if row['Exited'] == 0:
        is_in_pause = pd.notna(row['pause_start']) and (row['month'] >= row['pause_start']) and (row['month'] <= row['pause_end'])
        
        if is_in_pause:
            return 0.0  # Si está en pausa, la actividad de este mes es CERO.
        else:
            return lam  # Si no, tiene su actividad normal.
    # ==========================================================================
            
    # --- LÓGICA PARA CHURNERS (Sin cambios) ---
    else: # (row['Exited'] == 1)
        exit_date = pd.to_datetime(row['ExitDate'])
        month_end = row['month'] + pd.offsets.MonthEnd(0)
        
        if month_end >= exit_date:
            return 0.0
        
        is_repentino = client_repentino_map.get(row['CustomerId'], False)
        if is_repentino:
            return lam
        else:
            months_before = int(((exit_date.to_period('M') - row['month'].to_period('M')).n))
            if months_before <= last_n_months_decay and months_before > 0:
                frac = months_before / float(last_n_months_decay)
                base_decay_factor = (decay_final_frac + (1 - decay_final_frac) * frac)
                noise_decay_factor = np.clip(RNG.normal(loc=1.0, scale=DECAY_NOISE_SIGMA), 0.5, 1.5)
                lam = lam * base_decay_factor * noise_decay_factor
                return max(0.01, lam)
    
    return lam # Devuelve el lambda para los meses de churners fuera de la ventana de decaimiento


# --- APLICACIÓN DE LA FUNCIÓN Y GENERACIÓN DE DATOS (SIN CAMBIOS) ---

# Aplicar la función mejorada para calcular la actividad mensual de cada cliente.
grid['lambda'] = grid.apply(compute_lambda_row, axis=1)

# 3) Samplear el número de sesiones por cliente-mes usando una distribución de Poisson.
grid['n_sessions'] = RNG.poisson(grid['lambda'].values)

# Eliminar meses sin actividad para optimizar memoria.
grid = grid[grid['n_sessions'] > 0].reset_index(drop=True)

# 4) Expandir a nivel de sesión.
rep_counts = grid['n_sessions'].values
total_sessions = rep_counts.sum()
print(f"Total sesiones generadas: {total_sessions:,}")

session_customer = np.repeat(grid['CustomerId'].values, rep_counts)
session_month = np.repeat(grid['month'].values.astype('datetime64[ns]'), rep_counts)

# 5) Generar propiedades para cada sesión.
cust_to_cluster = clients.set_index('CustomerId')['CustomerSegment'].to_dict()
session_cluster = np.vectorize(cust_to_cluster.get)(session_customer)

# Duración de la sesión.
mu_map = {k: v['duration_mu'] for k, v in cluster_params.items()}
sigma_map = {k: v['duration_sigma'] for k, v in cluster_params.items()}
mu_arr = np.vectorize(mu_map.get)(session_cluster)
sigma_arr = np.vectorize(sigma_map.get)(session_cluster)
durations = np.clip(RNG.lognormal(mean=mu_arr, sigma=sigma_arr), 0.5, 180.0)

# Probabilidades base de cada evento.
p_transfer_base = np.vectorize(lambda cl: cluster_params[cl]['p_transfer'])(session_cluster)
p_payment_base  = np.vectorize(lambda cl: cluster_params[cl]['p_payment'])(session_cluster)
p_invest_base   = np.vectorize(lambda cl: cluster_params[cl]['p_invest'])(session_cluster)
p_push_base     = np.vectorize(lambda cl: cluster_params[cl]['p_push_open'])(session_cluster)
p_fail_base     = np.vectorize(lambda cl: cluster_params[cl]['p_failed_login'])(session_cluster)

# Inyectar ruido a las probabilidades.
noise_factors = np.clip(RNG.normal(loc=1.0, scale=PROB_NOISE_SIGMA, size=total_sessions), 0.5, 1.5)
p_transfer = np.clip(p_transfer_base * noise_factors, 0.001, 0.999)
p_payment  = np.clip(p_payment_base * noise_factors, 0.001, 0.999)
p_invest   = np.clip(p_invest_base * noise_factors, 0.001, 0.999)
p_push     = np.clip(p_push_base * noise_factors, 0.001, 0.999)
p_fail     = np.clip(p_fail_base * noise_factors, 0.001, 0.999)

# Muestreo de eventos para cada sesión.
used_transfer = RNG.random(total_sessions) < p_transfer
used_payment  = RNG.random(total_sessions) < p_payment
used_invest   = RNG.random(total_sessions) < p_invest
opened_push   = RNG.random(total_sessions) < p_push
failed_login  = RNG.random(total_sessions) < p_fail

# Asignar fechas aleatorias dentro de cada mes.
month_starts = pd.to_datetime(session_month).to_period('M').to_timestamp()
month_ends = (pd.to_datetime(session_month).to_period('M') + 1).to_timestamp() - pd.Timedelta(days=1)
days_in_month = (month_ends - month_starts).days + 1
rand_days = RNG.integers(0, days_in_month, size=total_sessions)
session_dates = month_starts + pd.to_timedelta(rand_days, unit='D')

# Crear IDs de sesión.
session_ids = np.arange(total_sessions).astype(str)

# Ensamblar el DataFrame final a nivel de sesión.
df_sessions_synth = pd.DataFrame({
    'SessionId': session_ids,
    'CustomerId': session_customer,
    'SessionDate': session_dates,
    'DurationMin': durations,
    'UsedTransfer': used_transfer.astype(int),
    'UsedPayment': used_payment.astype(int),
    'UsedInvest': used_invest.astype(int),
    'OpenedPush': opened_push.astype(int),
    'FailedLogin': failed_login.astype(int)
})

# 6) Agregados por cliente a nivel mensual.
df_sessions_synth['month'] = df_sessions_synth['SessionDate'].dt.to_period('M').dt.to_timestamp()


# 8) Validaciones y resumen final.
print("\n--- Resumen de Datos Generados ---")
print(f"Sesiones totales: {len(df_sessions_synth):,}")

Total sesiones generadas: 1,028,596

--- Resumen de Datos Generados ---
Sesiones totales: 1,028,596


In [66]:
# NOTA: Asegúrate de que df_sessions_synth y df_clients estén disponibles en este punto.

# 1. Calcular la fecha de la última sesión por cliente en df_sessions_synth
# La columna de fecha en df_sessions_synth es 'date'.
last_session_date = df_sessions_synth.groupby('CustomerId')['SessionDate'].max().reset_index(name='LastSessionDate')

# 2. Filtrar clientes que abandonaron y unirlos con su última sesión
df_churned = df_clients[df_clients['Exited'] == 1].copy()
# Seleccionamos las columnas relevantes y realizamos el merge
df_churned_check = df_churned[['CustomerId', 'ExitDate', 'Exited']].merge(
    last_session_date, on='CustomerId', how='left'
)

# 3. Realizar la comprobación de integridad
# El error ocurre si la ExitDate es ANTES de la LastSessionDate.
df_churned_check['TemporalIntegrityError'] = df_churned_check['ExitDate'] < df_churned_check['LastSessionDate']

# 4. Reportar los casos problemáticos (donde el error es True)
integrity_violations = df_churned_check[df_churned_check['TemporalIntegrityError'] == True]

print(f"Clientes que abandonaron ('Exited'=1): {df_churned.shape[0]}")
print(f"Clientes con violación de integridad temporal: {integrity_violations.shape[0]}")

if integrity_violations.empty:
    print("\n✅ ¡Validación completada! No se encontraron sesiones después de las fechas de salida registradas.")
else:
    print("\n🚨 ¡ATENCIÓN! Se encontraron violaciones de integridad. Primeros 5 casos:")
    # Usamos .dt.date para imprimir las fechas de forma limpia
    violations_clean = integrity_violations[['CustomerId', 'ExitDate', 'LastSessionDate']].copy()
    violations_clean['ExitDate'] = violations_clean['ExitDate'].dt.date
    violations_clean['LastSessionDate'] = violations_clean['LastSessionDate'].dt.date
    
    print(violations_clean.head())
    print("\nEstas filas indican que el cliente tuvo actividad de sesión posterior a su fecha de salida ('ExitDate').")

Clientes que abandonaron ('Exited'=1): 2037
Clientes con violación de integridad temporal: 0

✅ ¡Validación completada! No se encontraron sesiones después de las fechas de salida registradas.


In [67]:
df_sessions_synth.sort_values(by='SessionDate', inplace=True)

In [68]:
df_sessions_synth

,SessionId,CustomerId,SessionDate,DurationMin,UsedTransfer,UsedPayment,UsedInvest,OpenedPush,FailedLogin,month
435120,435120,15782418,2024-01-01,10.022934,0,0,1,0,0,2024-01-01
789177,789177,15699294,2024-01-01,7.582242,0,0,0,0,0,2024-01-01
867162,867162,15706799,2024-01-01,8.959697,1,0,0,1,0,2024-01-01
225881,225881,15761763,2024-01-01,56.510289,0,0,0,0,0,2024-01-01
42552,42552,15801559,2024-01-01,77.436571,0,0,0,0,0,2024-01-01
...,...,...,...,...,...,...,...,...,...,...
37937,37937,15594720,2025-12-31,24.989817,1,0,0,0,0,2025-12-01
1002978,1002978,15705174,2025-12-31,17.901046,1,0,0,0,0,2025-12-01
1002982,1002982,15705174,2025-12-31,2.478904,0,0,0,1,0,2025-12-01
397884,397884,15694450,2025-12-31,15.121201,0,0,0,0,0,2025-12-01


# <font color="#6e8ebc">**Guardado Datasets**</font>

In [69]:
df_transactions_synth.columns

Index(['TransactionId', 'CustomerId', 'TransactionDate', 'Amount',
       'TransactionType'],
      dtype='object')

In [70]:
df_transactions_synth.info()

<class 'pandas.core.frame.DataFrame'>
Index: 1939956 entries, 284065 to 250577
Data columns (total 5 columns):
 #   Column           Dtype         
---  ------           -----         
 0   TransactionId    object        
 1   CustomerId       int64         
 2   TransactionDate  datetime64[ns]
 3   Amount           float64       
 4   TransactionType  object        
dtypes: datetime64[ns](1), float64(1), int64(1), object(2)
memory usage: 88.8+ MB


In [71]:
df_sessions_synth.columns

Index(['SessionId', 'CustomerId', 'SessionDate', 'DurationMin', 'UsedTransfer',
       'UsedPayment', 'UsedInvest', 'OpenedPush', 'FailedLogin', 'month'],
      dtype='object')

In [72]:
df_clients.head()

,RowNumber,CustomerId,Surname,CreditScore,Geography,Gender,Age,Tenure,Balance,NumOfProducts,HasCrCard,IsActiveMember,EstimatedSalary,Exited,ExitDate,cluster_time,wealth_score,cluster_wealth,engagement_score,cluster_engagement,wealth_score_norm,engagement_score_norm,cluster_time_num,final_score,cluster_final,CustomerSegment
0,1,15634602,Hargrave,619,France,Female,42,2,0.00,1,1,1,101348.88,1,2025-11-15,Medium,40601.452,Low,1.0,Low,0.220009,0.25,0.5,0.288004,2,Cliente potencial
1,2,15647311,Hill,608,Spain,Female,41,1,83807.86,1,0,1,112542.58,0,NaT,Medium,86981.762,Medium,0.8,Low,0.471943,0.15,0.5,0.348777,4,Standard
2,3,15619304,Onio,502,France,Female,42,8,159660.80,3,1,0,113931.57,1,2025-12-28,High,125453.228,High,1.7,High,0.680918,0.60,1.0,0.712367,3,VIP
3,4,15701354,Boni,699,France,Female,39,1,0.00,2,0,0,93826.63,0,NaT,Low,37600.552,Low,1.0,Low,0.203708,0.25,0.0,0.181483,1,Poco Valor
4,5,15737888,Mitchell,850,Spain,Female,43,2,125510.82,1,1,1,79084.10,0,NaT,Medium,94474.050,Medium,1.0,Low,0.512641,0.25,0.5,0.405056,4,Standard


In [73]:
df_clients.columns

Index(['RowNumber', 'CustomerId', 'Surname', 'CreditScore', 'Geography',
       'Gender', 'Age', 'Tenure', 'Balance', 'NumOfProducts', 'HasCrCard',
       'IsActiveMember', 'EstimatedSalary', 'Exited', 'ExitDate',
       'cluster_time', 'wealth_score', 'cluster_wealth', 'engagement_score',
       'cluster_engagement', 'wealth_score_norm', 'engagement_score_norm',
       'cluster_time_num', 'final_score', 'cluster_final', 'CustomerSegment'],
      dtype='object')

In [74]:
df_clients = df_clients.drop(['cluster_time', 'wealth_score', 'cluster_wealth', 'engagement_score',
       'cluster_engagement', 'wealth_score_norm', 'engagement_score_norm',
       'cluster_time_num', 'final_score', 'cluster_final'], axis=1)

In [75]:
df_transactions_synth['TransactionDate'].min()

Timestamp('2024-01-01 00:00:00')

In [76]:
df_sessions_synth['SessionDate'].min()

Timestamp('2024-01-01 00:00:00')

In [77]:
df_transactions_synth = df_transactions_synth[df_transactions_synth['TransactionDate'] > '2024-12-31']
df_sessions_synth = df_sessions_synth[df_sessions_synth['SessionDate'] > '2024-12-31']

In [78]:
df_clients.to_parquet(DATA_PATH/'clients.parquet', index=False)
df_transactions_synth.to_parquet(DATA_PATH/'transactions.parquet', index=False)
df_sessions_synth.to_parquet(DATA_PATH/'app_sessions.parquet', index=False)